pallets/reward-distribution/src/lib.rs

```diff
@@ -99,5 +155,144 @@ pub mod pallet {
 			Self::deposit_event(Event::<T>::RewardPerBlockAdapted(new_reward_per_block));
 			Ok(())
 		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::collect_reward())]
+		#[transactional]
+		pub fn collect_reward(
+			origin: OriginFor<T>,
+			vault_id: DefaultVaultId<T>,
+			reward_currency_id: CurrencyId<T>,
+			index: Option<T::Index>,
+		) -> DispatchResult {
+			//distribute reward from reward pool to staking pallet store
+			let caller = ensure_signed(origin)?;
+			let reward = ext::pooled_rewards::withdraw_reward::<T>(
+				&vault_id.collateral_currency(),
+				&vault_id,
+				reward_currency_id,
+			)?;
+			ext::staking::distribute_reward::<T>(&vault_id, reward, reward_currency_id)?;
+
+			//withdraw the reward for specific nominator
+			let rewards =
+				ext::staking::withdraw_reward::<T>(&vault_id, &caller, index, reward_currency_id)?;
+
+			if rewards == (BalanceOf::<T>::zero()) {
+				return Err(Error::<T>::NoRewardsForAccount.into())
+			}
+
+			//either transfer from fee account if the reward is
+			//not the native currency, or just mint it if it is
+			let native_currency_id = T::GetNativeCurrencyId::get();
+			if reward_currency_id == native_currency_id {
+				T::Balances::deposit_creating(&caller, rewards);
```

Comment:
There is a potential flaw here when it comes to handling the native currency. The `collect_reward()` function resembles the `Fee::withdraw_rewards()` extrinsic, but it differs in the way that you can specify the `reward_currency_id` that you want to get paid out. 

However, these two extrinsics differ in another way: Here, we mint the rewards for the native currency, and in `Fee::withdraw_rewards()`, they are paid out from the reward pool. 
Now the problem is that, if a user tried to call `Fee::withdraw_rewards()`, the `Fee` pallet would not be able to pay out the whole reward of the user because it just does not have that many native tokens to transfer to them. Because they are not minted and just imaginary yet. 

In this `collect_reward` extrinsic we do mint these native tokens. But we neglect the native tokens that were collected into the fee pool. 

I would suggest that we remove the `Fee::withdraw_rewards()` extrinsic altogether, since the `collect_reward` extrinsic offers the same logic but better. 

Besides that, we should change this to first transfer all the native currency that is available in the fee pool and then pay out the rest with `deposit_creating`. Maybe something like this pseudo_code
```
let available_fees_native = T::Balances::balance_of(&Self::fee_pool_account_id(), native_currency_id);
let amount_fees_native: currency::Amount<T> = Amount::new(rewards, native_currency_id);
amount.transfer(&Self::fee_pool_account_id(), &caller)?;
let remaining_rewards = rewards.checked_sub(available_fees_native)?;
T::Balances::deposit_creating(&caller, remaining_rewards);
```

Reply:
Thanks @ebma this is a very good observation. Indeed the extrinsic on the fee pallet I left it only to discuss about it but also had the idea of removing it, but nevertheless I did not think about the difference in where the native tokens are coming from. 
Besides removing this extrinsic and it's corresponding code, I had the idea that it would never be the case that a fee is paid in native currency, since only wrapped currencies are the one issued or redeemed which end up as fees on the fee account. Is this correct?
If that case is correct, then we should always pay the native rewards per block by minting (deposit_creating), and we would not have the problem of dealing with native currency in the fee account.

Reply:
Right, good point. The griefing is also the only part that I was worried about. 
You can see [here](https://github.com/pendulum-chain/pendulum/blob/08b2eef8b43d216cc176571d75052dee22866b22/runtime/amplitude/src/lib.rs#L1165) that the currency we define for locking the griefing collateral is native.
However, in case this fee is indeed consumed because the user did not execute the issue request in time (ie. the user was griefing), the griefing fee is [transferred directly to the vault accont ](https://github.com/pendulum-chain/spacewalk/blob/8bb0d639c6f2e23126f72e39f56c6a497cf2e2d4/pallets/issue/src/lib.rs#L570)and not distributed to the fee pool. This behaviour also matches step 5) described in the spec [here](https://spec.interlay.io/spec/issue.html).

Soo... off the top of my hat I also don't see where we would collect fees in native. But we need to think carefully about that of course.

But on the other hand I don't see a problem if we mix the fee + reward handling for the native currency. We can always just try to exhaust the native balance that the fee pallet has of the native currency (if any, maybe there will never be any, we need to check this) and then mint the rest. The receiver of the reward does not need to care if they come from a fee pool or are minted here. The overall reward stays the same anyways.

Or am I missing something?

Reply:
No there is not a problem at all mixing, and I have to agree that the good thing is that only requires changes on the new pallet. So I will add logic that to handle cases we are perhaps overlooking.
On another related topic